File: Cdb/src/java/org/eso/ias/cdb/json/CdbFolders.java
package org.eso.ias.cdb.json;

import java.io.File;
import java.io.IOException;
import java.nio.file.Path;
import java.util.Objects;

/**
 * <P>The structure of the folders and the names of the files 
 * of the CDB are described in the 
 * <A href="https://github.com/IntegratedAlarmSystem-Group/ias/wiki/ConfigurationDatabase">CDB wiki</A>:
 * 
 * <pre>
* {@code
 * +CDB
 * 	| ias.json
 *  |+ Supervisor
 *     | <SupID-1>.json
 *     |
 *     | <SupID-n>.json
 *  |+ DASU
 *     | <DasuID-1>.json
 *     | ...
 *     | <DasuID-m>.json
 *  |+ ASCE
 *     | <AsceID-1>.json
 *     | ...
 *     | <AsceID-t>.json
 *  |+ IASIO
 *     | IASIO.json
 *     }
 * </pre>
 * 
 * Note that <code>CdbFolders</code> only deals with the folders of the CDB 
 * but not with the name of the files that ultimately depends on the selected format.
 * The example shown above is for JSON files.
 * 
 * @author acaproni
 *
 */
public enum CdbFolders {
	
	ROOT("CDB"),
	SUPERVISOR("Supervisor"),
	DASU("DASU"),
	ASCE("ASCE"),
	IASIO("IASIO");
	
	/**
	 *  The name of the folder
	 */
	public final String folderName;
	
	/**
	 * Constructor 
	 * 
	 * @param folderName The name of the folder in the CDB
	 */
	private CdbFolders(String folderName) {
		this.folderName=folderName;
	}
	
	/**
	 * Delete the subfolder, if it exists.
	 * 
	 * @param cdbParentPath: The path to the parent of the CDB
	 * @return <code>true</code> if the folder has been deleted,
	 * 	       <code>false</code> otherwise
	 */
	public boolean delete(Path cdbParentPath) {
		if (cdbParentPath==null) {
			throw new NullPointerException("The parent path of the CDB can't be null");
		}
		Path folderPath = buildPath(cdbParentPath);
		if (!exists(cdbParentPath)) {
			// The folder does not exist ==> nothing to do
			return false;
		}
		
		deleteFolder(folderPath.toFile());
		return true;
	}
	
	/**
	 * Recursively delete a folder from the file system.
	 * 
	 * @param folderToDelete
	 */
	private void deleteFolder(File folderToDelete) {
		Objects.requireNonNull(folderToDelete, "Cannot delete a NULL folder");
		
		File[] files = folderToDelete.listFiles();
	    if(files!=null) { //some JVMs return null for empty dirs
	        for(File f: files) {
	            if(f.isDirectory()) {
	                deleteFolder(f);
	            } else {
	                f.delete();
	            }
	        }
	    }
	    folderToDelete.delete();
	}
	
	/**
	 * Build the path of the subfolder
	 * 
	 * @param cdbParentPath: The path to the parent of the CDB
	 * @return The path to the subfolder
	 */
	private Path buildPath(Path cdbParentPath) {
		if (cdbParentPath==null) {
			throw new NullPointerException("The parent path of the CDB can't be null");
		}
		Path folderPath;
		if (this==ROOT) {
			folderPath=cdbParentPath.resolve(ROOT.folderName);
		} else {
			folderPath=cdbParentPath.resolve(ROOT.folderName).resolve(this.folderName);
		}
		return folderPath;
	}
	
	/**
	 * Get the path of the CDB folder and if it is the case, creates it.
	 * 
	 * @param cdbParentPath: The path to the parent of the CDB
	 * @param create: if <code>true</code> and the subfolder does not exist, then create it
	 * @return the path to the subfolder.
	 * @throws IOException If the folder is not writeable
	 */
	public Path getFolder(Path cdbParentPath, boolean create) throws IOException {
		if (cdbParentPath==null) {
			throw new NullPointerException("The parent path of the CDB can't be null");
		}
		Path folderPath=buildPath(cdbParentPath);
		File f = folderPath.toFile();
		if (exists(cdbParentPath)) {
			if (f.isDirectory() && f.canWrite()) {
				// No need to create the folder
				return folderPath;
			} else {
				throw new IOException(f.toString()+" exists but unusable: check permissions and type");
			}
		} else {
			f.mkdirs();
			return folderPath;
		}
	}
	
	/**
	 * Check if the subfolder exists
	 * 
	 * @param cdbParentPath The path of the folder
	 * @return <code>true</code> if the folder exists,
	 * 		   <code>false</code> otherwise.
	 */
	public boolean exists(Path cdbParentPath) {
		Path path = buildPath(cdbParentPath);
		File f = path.toFile();
		return f.exists();
	}
	
	/**
	 * Get the path of a CDB folder and if it is the case, creates the folder.
	 * 
	 * @param cdbParentPath: The path to the parent of the CDB
	 * @param folder: the CDB folder to create 
	 * @param create: if <code>true</code> and the subfolder does not exist, then create it
	 * @return the path to the subfolder.
	 * @throws IOException In case of error getting the path
	 */
	public static Path getSubfolder(Path cdbParentPath, CdbFolders folder, boolean create) throws IOException {
		if (folder==null) {
			throw new NullPointerException("CDB subfolder can't be null");
		}
		return folder.getFolder(cdbParentPath, create);
	}
	
	/**
	 * Create all the subfolders of the CDB.
	 * 
	 * If a subfolder already exists then nothing is done
	 * 
	 * @param cdbParentPath: The path to the parent of the CDB
	 * @throws IOException in case of error creating the folders
	 */
	public static void createFolders(Path cdbParentPath) throws IOException {
		for (CdbFolders folder: CdbFolders.values()) {
			folder.getFolder(cdbParentPath, true);
		}
	}
}

File: Cdb/src/main/python/IasCdb/CdbFolders.py
"""
Python implementation of CdbFolders.java
"""

import os
import shutil

from enum import Enum

class Folders(Enum):
    ROOT = 1
    SUPERVISOR = 2
    DASU = 3
    ASCE = 4
    IASIO = 5
    TF = 6
    TEMPLATE = 7
    PLUGIN = 8
    CLIENT =9

class CdbFolders:
    '''
    CdbFolder is the python translation of the java CdbFolder.java

    The differences are mostly due to the fact that python implemntation 
    of enumerated differs from java 
    '''

    ALL_FOLDERS = {
        Folders.ROOT:"CDB",
        Folders.SUPERVISOR:"SUPERVISOR",
        Folders.DASU:"DASU",
        Folders.ASCE:"ASCE",
        Folders.IASIO:"IASIO",
        Folders.TF:"TF",
        Folders.TEMPLATE:"TEMPLATE",
        Folders.PLUGIN:"PLUGIN",
        Folders.CLIENT:"CLIENT"
    }

    def __init__(self, folder: Folders, cdb_parent_path: str) -> None:
        '''
        Constructor

        Args:
            folder: the forlder of this object
            cdb_parent_path: the path of the folder containig the CDB
        '''
        self.folder = folder
        self.cdb_parent_path = cdb_parent_path
        if not self.cdb_parent_path:
            raise ValueError("The parent path can't be null/empty")
        if self.folder==Folders.ROOT:
            self.folder_path = os.path.join(cdb_parent_path, CdbFolders.ALL_FOLDERS[Folders.ROOT])
        else:
            self.folder_path = os.path.join(cdb_parent_path, CdbFolders.ALL_FOLDERS[Folders.ROOT], CdbFolders.ALL_FOLDERS[folder])
        if not os.path.isdir(self.cdb_parent_path):
            raise ValueError(f"CDB parent folder {self.cdb_parent_path} does not exist!")

    def delete(self) -> bool:
        """
        Delete the subfolder represented by this object, if it exists.
	    
        Returns:
	        True if the folder has been deleted, False otherwise
        """
        if not os.path.isdir(self.folder_path):
            # The folder does not exist ==> nothing to do
            return False
        
        self._delete_folder()
        return True

    def _delete_folder(self) -> None:
        """
        Recursively delete a folder from the file system.
        Args:
            folder_to_delete: The folder to delete
        """
        if not os.path.isdir(self.folder_path):
            # The folder does not exist ==> nothing to do
            return        
        for root, dirs, files in os.walk(self.folder_path, topdown=False):
            for file in files:
                os.remove(os.path.join(root, file))
            for dir in dirs:
                os.rmdir(os.path.join(root, dir))
        os.rmdir(self.folder_path)

    def get_folder(self, create: bool) -> str:
        """
        Get the path of the CDB sub folder and if it is the case, creates it.

        Args:
            create: if True and the subfolder does not exist, then create it
        Returns:
            the path to the subfolder.
        """
        # does not exist: create if requested
        if create and not os.path.isdir(self.folder_path):
            os.makedirs(self.folder_path)
        return self.folder_path

    def exists(self) -> bool:
        """
        Check if the subfolder exists

        Returns:
            True if the folder exists, False otherwise
        """
        return os.path.exists(self.folder_path)

    @staticmethod
    def get_subfolder(cdb_parent_path: str, folder: Folders, create: bool) -> str:
        """
        Get the path of a CDB folder and if it is the case, creates the folder

        Args:
            cdbParentPath: The path to the parent of the CDB
            folder: the CDB folder to create
            create: if True and the subfolder does not exist, then create it
        Returns:
            the path to the subfolder
        """
        if not folder:
            raise ValueError("CDB subfolder can't be None")
        if not cdb_parent_path:
            raise ValueError("CDB parent folder can't be None")
        cdb_folders = CdbFolders(folder, cdb_parent_path)
        return cdb_folders.get_folder(create)

    @staticmethod
    def create_folders(cdb_parent_path: str):
        """
        Create all the subfolders of the CDB.
        If a subfolder already exists then nothing is done

        Args:
            cdbParentPath: The path to the parent of the CDB
        """
        if not cdb_parent_path:
            raise ValueError("CDB parent folder can't be None")
        for folder in CdbFolders.ALL_FOLDERS:
            folder = CdbFolders(folder, cdb_parent_path)
            folder.get_folder(True)
